### PR TITLE
add cc build support

### PR DESCRIPTION
**Pull request by ryancinsight**
Hi, it would be nice to not require needing cmake, I only have ubuntu and windows to test needed flags.

### CODE REVIEW

**Comment by ryancinsight** [comment]
@indygreg when you have free moment would you mind testing this on your mac as a git patch, should also remove need for cmake on optional allocators

**Comment by mjp41** [comment]
I understand the wish to remove dependencies, but to do that through replication seems problematic from a maintenance point of view.  If the snmalloc project changes settings, these will need mirroring here.  Personally, I would not a fan of this change as is.

Is there some way the cmake settings could be automatically extracted as part of CI?  Can we check equivalence between the flags set with CMake and with this on a selection of platforms in CI?  I don't want to be tracing a bug in two years because the builds are subtly different between `snmalloc` and `snmalloc-rs`.

Could you make a stronger case for why `CMake` is too burdensome a dependency for your use case?

File: snmalloc-sys/build.rs [hunk 1]
```diff
@@ -1,62 +1,75 @@
...
+    build.flag_if_supported("/Ob2");
+    build.flag_if_supported("/DNDEBUG");
+    build.flag_if_supported("/EHsc");
+    build.flag_if_supported("/std:c++17");
+    build.flag_if_supported("-O3");
+    build.flag_if_supported("-Wc++17-extensions");
+    build.flag_if_supported("-std=c++1z");
+    build.flag_if_supported("-std=gnu++1z");
+    build.flag_if_supported("-mcx16");
+	build.flag_if_supported("-fno-exceptions");
```
[the remark this answers is not in the document]
Comment (ryancinsight) [reply snmalloc-sys/build.rs:20]: Haha, I'm actually surprised by this, I must of added in after doing a cargo fmt, I'll make sure it goes away on next iteration!

**Comment by ryancinsight** [comment]
> I understand the wish to remove dependencies, but to do that through replication seems problematic from a maintenance point of view.  If the snmalloc project changes settings, these will need mirroring here.  Personally, I would not a fan of this change as is.
> 
I fully expect some iterations on this but while replication can seem problematic I feel cc will provide us some finer control of the build systems and toolchains. For one, flag_if_supported allows us to put in the acceptable flags and automatically decide to use them if supported. There is also ways to check the toolchains and bias the build flags like cmake does but I kept simple for now and placed all flags found in cmakelist. 
> 
> Is there some way the cmake settings could be automatically extracted as part of CI?  Can we check equivalence between the flags set with CMake and with this on a selection of platforms in CI?  I don't want to be tracing a bug in two years because the builds are subtly different between `snmalloc` and `snmalloc-rs`.
> 
This is an interesting idea and will definitely look into it. Currently, I added all flags found in cmakelist but this may be redundant or unneeded, but if the compiler supports a flag in cmakelist than it should be used right now. I did add option for local dynamic tls as well though. I'll use --verbose --verbose flags on build the share cmake and cc flags on windows soon so we can compare.
> 
> Could you make a stronger case for why `CMake` is too burdensome a dependency for your use case?

The reason not to use cmake is that I've noticed it will mask a lot of warnings. I am routinely having to add libs manually because compiler ignores pragmas on some system for example. Further, pyoxidizer is mostly meant to package Python projects and I fully expect that good percentage of these individuals may not even have cmake installed. Cc just requires a c compiler to be available. Cc also tries to limit need for manually tracking adding libs, it creates a manifest while copying and does this for you, which is why I no longer needed to track build location.

**Comment by ryancinsight** [comment]
@mjp41 based on comments I decided to reimplement as a build feature so that cmake or cc can be used with cmake as default, making cmake-rs a download but should not be imported or used if build_cc is used as a feature, trying to figure out if there is a mutual exclusive option for features.

Cmake is using:
[snmalloc-sys 0.2.25]   C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Tools\MSVC\14.
16.27023\bin\HostX64\x64\CL.exe /c /I"D:\fork\snmalloc-rs\snmalloc-sys\snmalloc\src" /Zi /nologo /W4 /WX
 /diagnostics:classic /O2 /Ob2 /D NDEBUG /D "CMAKE_INTDIR=\"Release\"" /D _MBCS /Gm- /EHsc /MD /GS /fp:p
recise /Zc:wchar_t /Zc:forScope /Zc:inline /std:c++17 /Fo"snmallocshim-1mib-rust.dir\Release\\" /Fd"snma
llocshim-1mib-rust.dir\Release\snmallocshim-1mib-rust.pdb" /Gd /TP /wd4127 /wd4324 /wd4201 /errorReport:
queue  -Brepro "D:\fork\snmalloc-rs\snmalloc-sys\snmalloc\src\override\rust.cc"

while cc is using:
[snmalloc-sys 0.2.25] running: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\BuildTools\\VC\\
Tools\\MSVC\\14.16.27023\\bin\\HostX64\\x64\\cl.exe" "-nologo" "-MD" "-O2" "-Brepro" "-I" "snmalloc/src"
 "-W4" "/O2" "/Zi" "/nologo" "/W4" "/WX" "/wd4127" "/wd4324" "/wd4201" "/Ob2" "/DNDEBUG" "/EHsc" "/std:c
++17" "/Gd" "/TP" "/Gm-" "/GS" "/fp:precise" "/Zc:wchar_t" "/Zc:forScope" "/Zc:inline" "-FoD:\\fork\\snm
alloc-rs\\target\\x86_64-pc-windows-msvc\\release\\build\\snmalloc-sys-9814356dbc381ffd\\out\\snmalloc/s
rc/override/rust.o" "-c" "snmalloc/src/override/rust.cc"

File: snmalloc-sys/build.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ryancinsight) [reply snmalloc-sys/build.rs]: corrected!

**Comment by mjp41** [comment]
@ryancinsight thanks so much for making the changes.  I think this gives a good approach, if someone is experiencing problems with the `cc` feature, then they can try with `cmake`, and that gives us a quick elimination of possibilities, when debugging where the problem may lie.

**Comment by ryancinsight** [comment]
I added separate tests to use build_cc feature in rust build checks after 1mb test in order to help with debugging separately from cmake in future.

**Comment by SchrodingerZhu** [comment]
LGTM. @mjp41 any suggestion?

**Comment by SchrodingerZhu** [comment]
@ryancinsight would u mind add the feature flag description to README?

**Comment by ryancinsight** [comment]
> @ryancinsight would u mind add the feature flag description to README?

of course, I also wanted to point out the addition of dynamic local tls feature, that works for cc but not sure it will override the cmake version. Jemalloc and mimalloc had to add it for issues with "cannot allocate static memory block" so I didn't think it would hurt as an option.

**Review by SchrodingerZhu — APPROVED** [review]

**Comment by mjp41** [comment]
> LGTM. @mjp41 any suggestion?

LGTM as far as I understand the build system.

**Pull request merged**